### ISSUE TITLE
wncklet: Use cairo surfaces to render icons

### ISSUE THREAD

**Issue by vkareh** [mate-desktop/mate-panel#1003]
Hi @mate-desktop/core-team 

This is an issue to track this patch: https://gitlab.gnome.org/GNOME/libwnck/merge_requests/10

It replaces GdkPixbuf manipulation with the cairo_surface equivalents in libwnck. As a result, icons are now sharp in HiDPI displays.

Please test! (no need to rebuild mate-panel or anything, just restart panel after bulding libwnck)

**Workspace Switcher** (before/after)
![before-pager](https://user-images.githubusercontent.com/259780/61421299-7cb0b780-a8d4-11e9-814c-836d0cc46336.png)
![after-pager](https://user-images.githubusercontent.com/259780/61421295-7c182100-a8d4-11e9-9ede-04c7b6aab74c.png)

**Window Selector** (before/after)
![before-selector](https://user-images.githubusercontent.com/259780/61421300-7cb0b780-a8d4-11e9-8e89-ed8014ca3b3b.png)
![after-selector](https://user-images.githubusercontent.com/259780/61421297-7c182100-a8d4-11e9-89e8-1516e43e737c.png)

**Window List** (before/after)
![before-tasklist](https://user-images.githubusercontent.com/259780/61421301-7cb0b780-a8d4-11e9-861b-b2ecf3db31bc.png)
![after-tasklist](https://user-images.githubusercontent.com/259780/61421298-7c182100-a8d4-11e9-8aa6-eb771b1000fa.png)

**Comment by lukefromdc** [comment]
I too can confirm this works REALLY well for me with a 4K monitor

**Comment by raveit65** [comment]
This works nice here, fedora's libwnck manitainer is waiting to add this to libwnck3 rpm ;)

**Comment by lukefromdc** [comment]
We do get a double-scaling of the window menu icon in marco or compiz-reloaded with this as discussed in  gitlab.gnome.org/GNOME/libwnck/merge_requests/10 , only icons with a relative size are affected at least in compiz

**Comment by vkareh** [comment]
Yes, I'll address this early next week, wasn't my intention to change the wnck window API, so I'll revert that part.

**Comment by lukefromdc** [comment]
After about a week of work, I was able to write my own version of the libwnck PR that doesn't break icon sizes in marco or compiz-reloaded. Those icons still don't scale sharp but that needs to be fixed in marco and compiz, as it can't be helped from libwnck it seems. That can be done by using my new internal function wnck_window_get_mini_icon_scaled (it is exposed same as wnck_window_get_mini_icon) and not scaling in the decorator, using the existing PR as it stands and not rescaling the icon in the window decorator, or potentially by using wnck_window_get_icon_name to get just the name of the icon and then loading it directly. 

Anyway, I now have sharp panel tasklist icons and same behavior as before in compiz and marco.

See https://gitlab.gnome.org/GNOME/libwnck/merge_requests/10#note_569782
and
https://gitlab.gnome.org/lukefromdc/libwnck/tree/support-cairo-surface-icons

unfortunately I was unable to open a PR of my own due to gitlab issues

**Comment by vkareh** [comment]
Thanks for that, I can try to factor that in into my PR as soon as I have the chance. I had tried scaling the icons directly in the tasklist code and the taskbar ended up blurry :/

**Comment by lukefromdc** [comment]
I have found in general that scaling up (even in cairo) always gives blurry results and in the case of letting GTK do it automatically did even in the case of using .svg icons for window manager min/max/close icons. My guess is that the ways that work have the effect of forcing the icon to initially load from a larger size (in fact, I observed that directly in Caja when I had some mismatched icons in a larger size) instead of scaling up from a smaller size. When I tried using cairo_scale in libwnck in the taskbar myself, I too got blurry icons-and that after hours of experiments.

Thus I think we need to keep the old API around giving the same icon sizes as before, and also expose the new one for using scaled icons that are actually sharp instead of scaling up smaller ones in window managers etc. Using libwnck from my branch, I tried this change locally in my version (from my dev-all branch of compiz-reloaded) and it worked very nicely, giving sharp window icons. The code for the old API is commented out, could be behind version checks later:
```

/*
    d->icon_pixbuf = wnck_window_get_mini_icon (win);
	icon_width = gdk_pixbuf_get_width(d->icon_pixbuf);
    icon_height = gdk_pixbuf_get_height(d->icon_pixbuf);
    d->icon_pixbuf = gdk_pixbuf_scale_simple (d->icon_pixbuf, icon_width * scale,
                                              icon_height * scale, GDK_INTERP_BILINEAR);
*/
    d->icon_pixbuf = wnck_window_get_mini_icon_scaled (win);

```

**Comment by vkareh** [comment]
I've updated the merge request on https://gitlab.gnome.org/GNOME/libwnck/merge_requests/10
It's a huge change, but now the API doesn't change. I've split the whole thing into a few commits.
